### PR TITLE
Enable panel data processing for every model including crossvalidation

### PR DESCRIPTION
**Pull request by LeonieFreisinger**
## **Background**
- offer support for the processing of panel dataset. This affects the following steps:
> - **pre-processing**: data-specific pre-processing steps, particularly standard split and cross-validation split
> - **modeling**:  all models capable of fitting and predicting panel data including model-specific processing
> - **post-processing**: data-specific post-processing step
> - **evaluation**: evaluation metric calculation

## **Key Changes**

**adapt processing for panel dataset of the following models:**

_**reviewer hint Prio1: focus if the implementations are handled consistently for every model.  These functions you can find in the `models_xxx.py` files and in the `models/tuils.py`**_
_**reviewer hint Prio1: focus on implementation for LinearRegressionModel. Asses if the current implementation easily generalizes for other darts models**_

✅ NeuralProphetModel
✅ TorchProphetModel
✅ LinearRegressionModel
✅ NaiveModel
✅ SeasonalNaiveModel
❌ ProphetModel —> not able to process panel data, only n_forecast = 1 ([https://ksolarski.github.io/posts/btc_prophet/](https://ksolarski.github.io/posts/btc_prophet/)). Further --> raise Error if interval_width is input

adapt commonly for all models:
✅ prep_or_copy_df() on same level for CV and simple

**enhance model initialization of relevant models to global input**
✅ NeuralProphetModel
✅ TorchProphetModel
all other models do not support global modeling

✅ **split_df() processing panel dataset** 

**crossvalidation_df() processing panel dataset:**
✅ split options for global (local, intersect, global-time)
✅ test with df of multi-length ts and different start/ end dates
✅ test with overlap

**test normalization-related input**
only relevant for NP
✅ test for local normalization
✅ test for global normalization

**metric calculation over multiple IDs** 

**model helper functions for multivariate**
_**reviewer hint Prio1: asses if this structure makes sense. Especially if the passed input arguments are chosen logically**_
✅ maybe_extend_df
✅ maybe_drop_added_values_from_df
✅ _check_min_df_len

**test cases**
—> tested on ERCOT dataset 
—> tested on dyi panel dataset based on PEYTON_MANNING
✅ all models on panel data with local model
✅ NP model on panel data with global model
✅ Test panel data on cross-validation split (with 3 different split options)
✅ local and global normalization

## **Review Checklist** 
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, added docstrings and data types to function definitions.
- [x] I have added the necessary pytests, 

Please make sure to follow our best practices in the [Contributing guidelines](https://github.com/ourownstory/neural_prophet/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: tot/models/models_naive.py [hunk 1]
```diff
@@ -97,6 +91,11 @@ def predict(self, df: pd.DataFrame, df_historic: pd.DataFrame = None):
                 dataframe containing column ``ds``, ``y``, and optionally ``ID`` with data
             df_historic : pd.DataFrame
                 dataframe containing column ``ds``, ``y``, and optionally ``ID`` with historic data
+            received_ID_col : bool
+                whether the ID col was present
```
Comment (noxan) [inline tot/models/models_naive.py:95]:
```suggestion
                whether the ID column was present
```

File: tot/models/models_simple.py [hunk 2]
```diff
@@ -193,6 +235,10 @@ def predict(self, df: pd.DataFrame, df_historic: pd.DataFrame = None):
                 dataframe containing column ``ds``, ``y``, and optionally ``ID`` with data
             df_historic : pd.DataFrame
                 dataframe containing column ``ds``, ``y``, and optionally ``ID`` with historic data
+            received_ID_col : bool
```
Comment (noxan) [inline tot/models/models_simple.py:238]: would make readability better for me personally
```suggestion
            received_id_column : bool
```

File: tot/models/models_simple.py [hunk 3]
```diff
@@ -204,11 +250,15 @@ def predict(self, df: pd.DataFrame, df_historic: pd.DataFrame = None):
         _check_min_df_len(df=df, min_len=self.n_forecasts)
         if df_historic is not None:
             df = self.maybe_extend_df(df_historic, df)
-        df, received_ID_col, received_single_time_series, _ = prep_or_copy_df(df)
         fcst_df = _predict_darts_model(
-            df=df, model=self, n_req_past_obs=self.n_lags, n_req_future_obs=self.n_forecasts, retrain=False
+            df=df,
+            model=self,
+            n_req_past_obs=self.n_lags,
+            n_req_future_obs=self.n_forecasts,
```
Comment (noxan) [inline tot/models/models_simple.py:257]: I assume `obs` are `observations`? being explicit and verbose makes it easier to read for people less familiar with the code

**Review by noxan — APPROVED** [review]
LGTM overall, it is a lot of code to review so I mainly skimmed over it, especially the dataframe utilities are a lot - might make sense to add tests for them at some point in time and adding some comments in between your lines of code would help to read along.

File: tot/models/models_simple.py [hunk 4]
```diff
@@ -204,11 +250,15 @@ def predict(self, df: pd.DataFrame, df_historic: pd.DataFrame = None):
         _check_min_df_len(df=df, min_len=self.n_forecasts)
         if df_historic is not None:
             df = self.maybe_extend_df(df_historic, df)
```
Comment (ourownstory) [inline tot/models/models_simple.py:252]: Suggestion: pass args by name, not position, to avoid positional mistakes.

Comment (ourownstory) [reply tot/models/models_simple.py:252]: note: in `maybe_extend_df` and `maybe_drop_added_values_from_df` you could save a line each by not setting, but directly passing `samples = self.n_lags`

**Review by ourownstory — APPROVED** [review]
Please check carefully before merging:
* Do the values of `predicted` passed to `maybe_drop_added_values_from_df` match the `df` length, including the first `n_lags` (maybe added) inputs? Different models may give different prediction lengths - omitting the first inputs or returning their values as none. `drop_first_inputs_from_df` drops the first samples from both df and predicted. if there is a mismatch, predicitons and df datestamps may be misaligned.


Questions: 
* Can we vectorize `drop_first_inputs_from_df` operation over "ID" columns?
* should we separate darts and Prophet models out or `tot/models/models_simple.py`?
* do we need to nest `_predict_raw_darts_model` in  `_predict_darts_model` (it seems to me that  `_predict_raw_darts_model` is not used outside `_predict_darts_model`. Maybe we can even skip the conversion to dicts and back?

Please Check:
`_check_min_df_len` 
We are currently checking if the df is at least `n_forecasts` long. I am not sure if this is required when predicting? (edit: never mind, we want to only predict over areas where we have data so we can evaluate the model performance, so we want this to match.)
We currently do not check whether it is `n_lags` long (plus `n_forecasts`), (should be checked after `maybe_extend_df`), but this may be necessary for many models?


RE: `LinearRegressionModel`
I am confused about ` output_chunk_length` - what does it do? In the docstring you say it is different from `n_forecasts`?
`multi_models` docstring tells me that multiple models are not created by `output_chunk_length` - I assume that `output_chunk_length` creates multiple linear regressions in matrix form of shape `n_lags` x `output_chunk_lengtht`? What happens when `output_chunk_length` and `n_forecasts` are not the same? Maybe we need to set them to the same?
Whether this implementation scales well to other Darts models is hard to to tell for me. It looks well laid out to me. But the real test would be to add another simple darts model, like ETS or ARIMA. :)

File: tot/models/utils.py [hunk 5]
```diff
@@ -144,7 +113,17 @@ def convert_df_to_TimeSeries(df, value_cols, freq) -> TimeSeries:
             time series to be fitted or predicted
 
     """
+    # Receives df with single ID column
+    received_single_ts = len(df["ID"].unique()) == 1
+
+    if not received_single_ts:
+        # pivot the resulting DataFrame to convert multiple time series
+        df = df.pivot(index="ds", columns="ID", values="y").rename_axis(columns=None).reset_index()
```
[the remark this answers is not in the document]
Comment (LeonieFreisinger) [reply tot/models/utils.py:121]: @ourownstory valid consideration! Darts supports regressors. However, they will be passed separately to the `fit()` function as a `TimeSeries `object, like this
```
model.fit(
    [train_air_scaled, train_milk_scaled],
    past_covariates=[air_covs, milk_covs],
);
```
That means we would need to extract them from our original dataframe and treat them separately. Do you have an idea, how to handle that mapping in a most lean way?
